### ISSUE TITLE
Default ActivitySpec visualizations

### ISSUE THREAD

**Issue by avaidyam** [BIDMCDigitalPsychiatry/LAMP-platform#266]
Currently, all data in the Prevent tab is visualized using the same line chart representation that is hard-coded into the dashboard. Instead, the `ActivitySpec` definition should provide the default document transform query (i.e. JSONata/Vega) to use for charts of that activity type. (This could also extend to `scoring`, `streak`, and `indicator` bubble functions that power other UI features of the app.)

1. The `ActivitySpec` provides the default graphing function, which is a JSONata expression.
1. If the `Activity` overrides the graphing function, that is used instead (similar to `icon`, `description`, etc.)
1. The Prevent tab UI will substitute the correct Activity ID and invoke the graphing function on the server-side (or client-side if required). 
1. The graphing function will retrieve the necessary data and generate a Vega spec that the app UI will render. 
    - If a JSON table schema (i.e. array of equal length arrays of strings/numbers, representing a CSV file) is returned, the data is rendered as an interactive DataFrame.

**Comment by lukeoftheshire** [comment]
I think adding scoring, not only for surveys, but activities, in general to LAMP would greatly increase both user and researcher QOL.

My "dream" implementation for survey scoring would be something like:
1. Each question is assigned a unique id by the survey creator - e.g. 'phq9-1'. This doesn't have to be shown to a user, but it should be unique for each question. This would have the side effect of removing the current reliance on exactly matching long strings that can lead to human error.
2. Each question type has its own scoring method (e.g. list questions have a corresponding list of numerical scores). Theoretically, this could be overridden by a provided JSONata/other query (for example to score a textbox question that could be scored in a few ways). This should be an optional part of survey creation - unscored surveys 
3. By default, the score is calculated by summing all scores for each question, which covers the majority of cases. However, this could also be overwritten with a JSONata query that takes an array in the format:
```
[{'question_id':'score'},
{'question_id2':'score2'}]
```
operates on it, and returns a numerical score. (This could probably also be done w/ temporal slices as long as the id is preserved.

Ideally, this would run when the survey is submitted and the score would be stored in the static_data of each ActivityEvent. This means that the score is preserved even if the method is adjusted later.


Other activities could also have a default scoring system that could be overwritten. Ideally, this method could also store this score in the static_data of each ActivityEvent.

**Comment by avaidyam** [comment]
> Each question is assigned a unique id by the survey creator - e.g. 'phq9-1'. This doesn't have to be shown to a user, but it should be unique for each question. This would have the side effect of removing the current reliance on exactly matching long strings that can lead to human error.

Minor note about this point - this technically exists today. It's called the index of the temporal slice (which should be guaranteed to match the index of the question in the Activity settings array). Maybe we should iron out a plan for maintaining consistency there (i.e. version editing of Activities?) instead of complicating survey questions with more new IDs. 

_(Also, this whole mechanism that you've proposed seems to be specific to surveys only and not all activities?)_

**Comment by lukeoftheshire** [comment]
Yes - this mechanism I'm proposing is specific to surveys, which I think may need to be more flexible than other cognitive games. There really aren't that many ways to score, for example, a Balloon Risk Task (how many points did you get?, e.g.). Surveys can require more complex methods. That being said, if every activity has a high level of scoring flexibility available that's fine too - it should just be *at least* available for surveys.

If we can guarantee that the index of the temporal slice remains constant, that's fine and is, as you point out, effectively an id. Currently that is hard to ensure with 100% certainty since, again as you note, activities are freely editable. Additionally, two main features we are planning to add, question randomization and optional questions/alternate 'paths' through a survey, might complicate this.

For example, once question randomization is implemented, if we choose to use temporal slice index as a 'pseudo-id' we will need another way to represent the order the questions were presented in - or vice versa if the index corresponds to the order the questions were shown in.

**Comment by avaidyam** [comment]
> If we can guarantee that the index of the temporal slice remains constant, that's fine and is, as you point out, effectively an id.

✅

> Currently that is hard to ensure with 100% certainty since, again as you note, activities are freely editable.

I think this is the main concern we need to be addressing. It's come up in the past as well for numerous other reasons. 

> Additionally, two main features we are planning to add, question randomization and optional questions/alternate 'paths' through a survey, might complicate this.

Issue #385 covers randomization - but we can discuss the implementation of that feature over there. Issue #265 covers question branching, but as explained in that issue, it may be wise to separate that version of surveys out as `lamp.advanced_survey`, so we can redesign the implementation from scratch over there. 

> if we choose to use temporal slice index as a 'pseudo-id' we will need another way to represent the order the questions were presented in

Right, which is why I'm somewhat opposed to introducing a whole new requirement and class of identifiers for survey questions. _(It makes more sense for activities to have such an ID to allow easily mapping localized versions of surveys, though, as opposed to localized activity titles/descriptions and requiring activities like Surveys to have support for localized questions... which would take much more engineering effort to implement.)_

**Comment by dcurrey88** [comment]
re #518: In addition it would be useful to have read-only sections in surveys (where there is no question only markdown text).

**Comment by avaidyam** [comment]
@lukeoftheshire I do agree that these features collectively enable a significant increase in study/analysis velocity, but we can't make this a priority issue at this time when we have issues like #541, #542, #550, etc. 

Do you have a proposed implementation we can begin to target, so that at least when we do activate and begin work on this functionality everything is ironed out ahead of time?

**Comment by lukeoftheshire** [comment]
Okay, I understand - I just want to make sure this does not go by the wayside as I think (as you say too) it would enable a big increase in velocity to have this; ease-of-use too, honestly, for both us and collaborators.

I think we can probably even start by rolling out scoring somewhat piecemeal. For example, the most common question we would need to score is multiple choice, which is pretty easy - just have each option associated with a specific point value. Any unscored multiple choice or other type of question (for now) could be just scored as 0. The calculated score could be reported as part of the static data object, which I believe is where we have stored score data for other tests.

My reasoning for this is that nailing down ways to score every question type we have prior to implementing scoring is likely to bog us down (how do we score text questions, for instance?). I think we should not let perfect be the enemy of good -- we may have to cover the general, most frequent use cases first - having multiple choice scoring *alone* would cover many, many commonly used scales.

To sum up, I would be thrilled if we could target implementing (optional) scoring for multiple choice type questions and reporting the score as part of the static data in an ActivityEvent. Further down the road I have ideas for how to score more types of questions, but I think this would be a good place to start.
